### PR TITLE
cling: update to 0.9

### PR DESCRIPTION
**Pull request by thomasjm**
Changelog for this version: https://github.com/root-project/cling/releases/tag/v0.9.

This is currently a WIP, putting it up to ask for some help from Cling maintainers.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thomasjm** [comment]
Hello @vgvassilev, I saw you made the latest release of Cling and I was hoping to ask you for some packaging help so that we can bring Cling 0.9 to Nixpkgs :)

I bumped the Cling and Clang source versions and changed LLVM to version 9, but I'm currently running into some CMake issues. Any chance you could point me towards the proper arguments?

```shellSession
> nix build .#cling -L

...
cling-unwrapped> CMake Error at /nix/store/gzxv33q3ah7065d3x4qb0xdl2dgf1g41-llvm-9.0.1-dev/lib/cmake/llvm/AddLLVM.cmake:637 (add_custom_target):
cling-unwrapped>   add_custom_target cannot create target "install-clang-cpp" because another
cling-unwrapped>   target with the same name already exists.  The existing target is a custom
cling-unwrapped>   target created in source directory "/build/clang/tools/driver".  See
cling-unwrapped>   documentation for policy CMP0002 for more details.
cling-unwrapped> Call Stack (most recent call first):
cling-unwrapped>   cmake/modules/AddClang.cmake:117 (add_llvm_install_targets)
cling-unwrapped>   tools/clang-shlib/CMakeLists.txt:18 (add_clang_library)
```

**Comment by thomasjm** [comment]
Also pinging @Axel-Naumann!

**Comment by thomasjm** [comment]
Just tried that and no change unfortunately. 

I wanted to check that the basic approach hasn't changed, right? In the past we've started with the Clang source tree (from the relevant `cling-vXXX` tag), then downloaded the Cling source and put it under `tools/cling`. And then invoked CMake from there.

The last time we did this we *didn't* need a patched LLVM. Is that still the case? It's always been a bit confusing to me since to this day, `cpt.py` seems to be doing stuff with `llvm-cling-patches` in the `fetch-llvm` function...

**Comment by vgvassilev** [comment]
I would need more context for the cmake error. 

I feel we include something from some pre-installed on the system clang/llvm which confuses cmake.

Can you attach the result of `cmake --trace-expand ...`?

**Comment by veprbl** [comment]
> The last time we did this we _didn't_ need a patched LLVM. Is that still the case?

For ROOT the compilation fails without this one:
https://github.com/NixOS/nixpkgs/blob/3761ba2119fde0972cf4aac48d0b4ccc87c53ab1/pkgs/applications/science/misc/root/default.nix#L59-L67

**Comment by vgvassilev** [comment]
> The last time we did this we didn't need a patched LLVM. Is that still the case?

LLVM9 was released broken, we need to pick up a patch from our patches again...

**Comment by thomasjm** [comment]
> Can you attach the result of cmake --trace-expand ...?

Here you go: https://gist.github.com/thomasjm/5fc4d0c562fe04ac7c242e3d0d723ce7

> For ROOT the compilation fails without this one

:+1: I'll look into that after we get past CMake issues

> LLVM9 was released broken, we need to pick up a patch from our patches again...

If later LLVM9 was fixed then Nix likely picked it up already

**Comment by vgvassilev** [comment]
> > Can you attach the result of cmake --trace-expand ...?
> 
> Here you go: https://gist.github.com/thomasjm/5fc4d0c562fe04ac7c242e3d0d723ce7

Looks like we have deliberately avoided that problem https://github.com/root-project/root/blob/323c21416cea9a64080698c23a7e2f6f6d1808bf/interpreter/CMakeLists.txt#L450-L454 in https://github.com/root-project/root/commit/0cddef5d3562a89fe254e0036bb7d5ca8a5d34d2

Can you do the same?

> 
> > For ROOT the compilation fails without this one
> 
> 👍 I'll look into that after we get past CMake issues
> 
> > LLVM9 was released broken, we need to pick up a patch from our patches again...
> 
> If later LLVM9 was fixed then Nix likely picked it up already

I doubt it was ever fixed. We realized that when it was too late for patch releasing version 9.

**Comment by thomasjm** [comment]
> Can you do the same?

I can try! The PR you sent scares me a little because it looks like ROOT has its own copy of Cling checked in, and these fixes are being made directly against that. Would it make sense to being such fixes over to Cling's repo at https://github.com/root-project/cling ?

**Comment by thomasjm** [comment]
As I work through this I have one question about the LLVM 9 patches: do we need the patched LLVM9 at *runtime* or only at compile time? I ask because when we packaged Cling 0.7, we were able to use the vanilla LLVM 5. 

I saw in [this change](https://github.com/root-project/root/commit/0cddef5d3562a89fe254e0036bb7d5ca8a5d34d2#diff-e181029772a3e9aeffcda7614c4b0b4af48d409768e3c48b8077f245a78f38c9R348-R372) that we need to point Cling to some modified LLVM header files. Seems like it's just because a certain class member needs to be changed from private to protected. So is that all we need to do, or do we also need to compile the patched LLVM?

**Comment by RaitoBezarius** [comment]
Is there anything I can do to help this PR? I am interested into dropping LLVM5 and Cling is probably the only obstacle on path at the moment.

**Comment by vgvassilev** [comment]
> As I work through this I have one question about the LLVM 9 patches: do we need the patched LLVM9 at _runtime_ or only at compile time? I ask because when we packaged Cling 0.7, we were able to use the vanilla LLVM 5.
> 
> I saw in [this change](https://github.com/root-project/root/commit/0cddef5d3562a89fe254e0036bb7d5ca8a5d34d2#diff-e181029772a3e9aeffcda7614c4b0b4af48d409768e3c48b8077f245a78f38c9R348-R372) that we need to point Cling to some modified LLVM header files. Seems like it's just because a certain class member needs to be changed from private to protected. So is that all we need to do, or do we also need to compile the patched LLVM?

I believe you don't need to recompile just for that. Maybe we can somehow patch that only file through the preprocessor or sed?

**Comment by thomasjm** [comment]
@vgvassilev thanks -- I just started looking at this again. This PR is currently trying to build the `v0.9` release, which is from May 2021. I see there have been a number of commits since then. Should we be trying to use a later commit? Is Cling itself going to have a new release soon?

**Comment by vgvassilev** [comment]
We will release as soon as we fix cling's current test failures due to the migration to llvm13. I cannot give a timeline unfortunately.

**Comment by thomasjm** [comment]
Okay, I actually got it to build! 

But it shows a weird error and fails to start, can you help with this @vgvassilev ?

```
: CommandLine Error: Option 'disable-symbolication' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
```

**Comment by vgvassilev** [comment]
This means that we have linked LLVMSupport.a twice somehow. Perhaps setting a breakpoint at the static initialization there should hint us who brings the second copy of this library.

**Comment by thomasjm** [comment]
FWIW, I wondered if this was because of how I made a copy of the include directory in order to patch it. I made a separate branch [here](https://github.com/NixOS/nixpkgs/compare/master...codedownio:nixpkgs:cling-0.9-patched-llvm?expand=1) to try patching LLVM directly (which causes it to be recompiled) and got the same result.

> Perhaps setting a breakpoint at the static initialization there should hint us who brings the second copy of this library.

Sure, I can try that -- could you elaborate on how to go about doing that? I assume you mean using `gdb` but I'm not sure what symbol to use.

**Comment by veprbl** [comment]
> ```
> : CommandLine Error: Option 'disable-symbolication' registered more than once!
> LLVM ERROR: inconsistency in registered CommandLine options
> ```

I think this could be related to an issue https://github.com/root-project/root/issues/12156 that we see when upgrading to ROOT 6.28  in #215187. You should be seeing errors about duplicate symbols from the linker.

**Comment by thomasjm** [comment]
Thanks @veprbl, I'm reading through those linked issues. Interesting that the ROOT 6.28 PR also removed those LLVM patches you mentioned earlier in the thread. I don't think I've seen any messages about duplicate symbols.

Anyway, it works now! I followed that rabbit hole until I found https://github.com/conda-forge/cling-feedstock/commit/f42b723d69bb8f88429c85fdfdff1a55460e92b9 and applied the same fix here.

Strangely, the same fix seems to have been applied to Cling already (or at least the version of Cling that lives in ROOT): https://github.com/Axel-Naumann/root/commit/655d22df2e5613801c300a1ebf9cef967bf288a0

**Comment by RaitoBezarius** [comment]
Hey, awesome you are making progress! Could I ask you also while on the subject to format your commits according to the CONTRIBUTING guide? Is there anything left to do on this PR?

**Comment by thomasjm** [comment]
In the past I've squashed and written the final commit message just before merging @RaitoBezarius. Not sure it would make sense to reformat them quite yet.

I think things are basically working. I was trying to get the new version working with the [xeus-cling](https://github.com/jupyter-xeus/xeus-cling) Jupyter kernel, but ran into problems. I'm not sure if the problems are on the Cling side though. 

I'd invite others to try out the branch at this point (any interest @veprbl?)

**Comment by RaitoBezarius** [comment]
> In the past I've squashed and written the final commit message just before merging @RaitoBezarius. Not sure it would make sense to reformat them quite yet.

You're right but it seems that stuff like adding the Jupyter part is not directly related to cling: 0.x -> 0.9. It was mostly to say "all changes that could look separate/atomic commits should be reformatted", but no problem. :)

> 
> I think things are basically working. I was trying to get the new version working with the [xeus-cling](https://github.com/jupyter-xeus/xeus-cling) Jupyter kernel, but ran into problems. I'm not sure if the problems are on the Cling side though.
> 
> I'd invite others to try out the branch at this point (any interest @veprbl?)

Got it!

**Comment by RaitoBezarius** [comment]
Hey, what's the state of this PR?

**Comment by thomasjm** [comment]
Well, I'm not having much luck getting other testers :P

I'd love for this to make the next Nixpkgs release though. Let me try again to get the Jupyter kernel stuff working--if that works well, I think it'll be good evidence that this upgrade is ready to go.

**Comment by veprbl** [comment]
@GrahamcOfBorg build cling

**Comment by veprbl** [comment]
Broken on darwin
```
       > -- Installing: /nix/store/qjkprp5845bh72y8chqya7vg9mds7kqf-cling-unwrapped-0.9/lib/cmake/cling/ClingTargets-release.cmake
       > -- Installing: /nix/store/qjkprp5845bh72y8chqya7vg9mds7kqf-cling-unwrapped-0.9/lib/cmake/cling/ClingConfig.cmake
       > -- Installing: /nix/store/qjkprp5845bh72y8chqya7vg9mds7kqf-cling-unwrapped-0.9/lib/cmake/clang/ClangTargets.cmake
       > -- Installing: /nix/store/qjkprp5845bh72y8chqya7vg9mds7kqf-cling-unwrapped-0.9/lib/cmake/clang/ClangTargets-release.cmake
       > -- Installing: /nix/store/qjkprp5845bh72y8chqya7vg9mds7kqf-cling-unwrapped-0.9/lib/cmake/clang/ClangConfig.cmake
       > -- Installing: /nix/store/qjkprp5845bh72y8chqya7vg9mds7kqf-cling-unwrapped-0.9/bin/hmaptool
       > cp: cannot stat '/build/clang/tools/cling/tools/Jupyter/kernel': No such file or directory
       > /nix/store/827pva44j6136ja5hqwyjh88wrwjqzw5-stdenv-darwin/setup: line 145: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/49fw8sk6yx2n9m531gjykqdram3mwdrr-cling-unwrapped-0.9.drv'.
error: 1 dependencies of derivation '/nix/store/x116hv8pzhfpfp9xcj6sbnkswg3mn3i8-cling-0.9.drv' failed to build
```
Please, fix or mark broken for it.

**Review by veprbl — APPROVED** [review]

**Comment by RaitoBezarius** [comment]
Urgh, I should not have merged this.
